### DISCUSSION TITLE
Points and Collision Mask editor feature requests

### DISCUSSION THREAD

**Discussion by AlexandreSi** [4ian/GDevelop#3876] [Enhancement ideas (and their technical discussions)]
Suggestions by @Silver-Streak:

1. Concave hull slicing into multiple convex shapes for collision masks. (Complicated and likely needs a separate story. I'll write up a request... eventually 😆 )
2. The ability to double click on a vertice to add a new point right at the double click location for masks, or a new custom point for poinys. This is pretty common when I was testing other engines for adventure game template ideas.

_Originally posted by @Silver-Streak in https://github.com/4ian/GDevelop/issues/3872#issuecomment-1119333298_

**Comment by Silver-Streak** [comment]
Thanks for bringing this over. Making two replies to allow easier tracking of responses to each idea.

New points:

- For adding new custom points, the idea would be to allow the user to double-click anywhere (even outside the sprite) in the points editor to generate a new point. 
- For adding new collision mask points/new vertexes, the idea would be to allow the user to   double-click anywhere on an existing vertice to add a new point. Require it being on an existing vertice would hopefully help avoid having to figure out what the order is on the list of vertices.

(Maybe these should be double right-click, incase people are concerned of accidental missed-double clicks.)

**Reply by tristanbob** [reply]
I think a double-click should be required.  Imagine trying to move a point, but you accidentally create a new one instead!  I think would happen quite often.

I'm not sure why you said "right-click", but right-click doesn't feel like a natural method for creating something.

**Reply by Silver-Streak** [reply]
Double right-click was mentioned specifically to avoid what you were worried about. 😄 

 Some people may be trigger happy with their left mouse button (or have conditions like arthritis or Parkinson's which can cause their fingers to shake/accidentally double clicks), having it on right mouse means they can avoid accidentally creating points when they were trying to grab an existing one.

Edit: I've edited my post above, as I can see it having been interpreted as a request for either a single left click or double right click.

**Comment by Silver-Streak** [comment]
For concave mask handling (Concave polygon decomposition), there's actually quite a few libraries/algorithms on this:
- https://github.com/schteppe/poly-decomp.js
- https://github.com/Louis-T/fernandez-polygon-decomposition

These were the two that popped up the most for me when I was digging into it originally.

I think the biggest thing would either be a manual button for the user to click when the "Polygon is not convex" warning comes up _or_ automatically doing it once they have a convex polygon and they click OK to save their shape. 

The biggest issue is that right now the engine (or at least raycasting in the engine) is detecting multiple collision masks on the same object as multiple objects (reference this issue opened here: https://github.com/4ian/GDevelop/issues/3563 ), so that would likely need to be fixed to avoid unexpected behavior by users.

**Reply by tristanbob** [reply]
This might be a good time to bring up this idea: https://github.com/4ian/GDevelop/discussions/3650

Instead of having users "draw" the collision mask, this feature would shrink the default AABB collision mask into a convex hull around the non-transparent pixels.
